main/src/main/resources/report.html

```diff
@@ -0,0 +1,229 @@
+            background: #ddd;
+        }
+        table.accordion > tbody > tr.notice.open {
+            background: #ddd;
+            color: black;
+        }
+        table.accordion > tbody > tr.description {
+            display: none;
+        }
+        table.accordion > tbody > tr.description.open {
+            display: table-row;
+        }
+
+        .desc-content {
+            padding: 0.5em;
+            border-bottom: 5px solid #000;
+            border-top: 5px solid #000
+        }
+        .desc-content h3 {
+            margin-top: 0;
+        }
+    </style>
+</head>
+
+<body>
+<h1>VALIDATION REPORT</h1>
+<h2><span th:text="${summary.noticeCount}">X</span> notices reported
+    (<span th:text="${summary.errorCount}">0</span> errors,
+    <span th:text="${summary.warningCount}">0</span> warnings,
+    <span th:text="${summary.infoCount}">0</span> infos)</h2>
+<p>Please visit <a href="https://github.com/MobilityData/gtfs-validator/blob/master/docs/NOTICES.md">NOTICES.md</a> and <a href="https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md">RULES.md</a>.</p>
+<table class="accordion">
+    <thead>
+    <tr>
+    <th>Notice Code</th><th>Notice Name</th><th>Severity</th><th>Total</th>
+    </tr>
+    </thead>
+    <tbody>
+    <span th:each="noticesBySeverityLevel: ${summary.noticesMap}">
+        <span th:each="noticeByCode: ${noticesBySeverityLevel.value}">
+            <tr class="notice">
+                <td th:text="${noticeByCode.key}" />
+                <td th:text="${noticeByCode.value.get(0).getName()}" />
+                <td th:class="${noticesBySeverityLevel.key.toString().toLowerCase()}" th:text="${noticesBySeverityLevel.key}" />
+                <td th:text="${noticeByCode.value.size()}" />
+            </tr>
+            <tr class="description">
+                <td colspan="4">
+                    <div class="desc-content">
+                        <h3 th:text="${noticeByCode.value.get(0).getName()}" />
+                        <p> Please find this notice description <a th:href="@{'https://github.com/MobilityData/gtfs-validator/blob/master/RULES.md#' + ${noticeByCode.value.get(0).getName()}}">here</a>.</p>
+                        <table>
+                            <thead>
+                                <tr>
+                                    <th:block th:each="field: ${noticeByCode.value.get(0).getFields()}" >
+                                        <th th:text="${field.getName()}" />
+                                    </th:block>
+                                </tr>
+                            </thead>
+                            <span th:each="notice: ${noticeByCode.value}">
+                                <tbody>
+                                <tr>
+                                <th:block th:each="field: ${notice.getFields()}">
+                                        <td th:text="${notice.getContext().getAsJsonObject().get(field.getName())}" />
```

Comment:
I have to admit I'm not a hug fan of the approach here of exposing a reflected Field[] list on each Notice and then looking up values via the serialized JSON object.  Java field reflection may not always be equivalent to JSON serialization (e.g. maybe someone marks that a field shouldn't be serialized in JSON?).  It also repeatedly serializes the same object as json in this loop.

I'd propose an alternative approach where you introduce a presentation model object, which is a wrapper around a Notice, that provides the specific methods needed for display.  This is the spirit behind ReportSummary, which is a presentation-level model object for NoticeContainer.

Such a Notice wrapper could call JSON serialization once on construction and use that representation to provide method for display.

Thoughts?

Reply:
I initially used `notice.context.getAsJsonObject().entrySet()` and looped over the entries to get the keys and values, which was working well, but I wasn't sure about the approach.

> I'd propose an alternative approach where you introduce a presentation model object, which is a wrapper around a Notice, that provides the specific methods needed for display. This is the spirit behind ReportSummary, which is a presentation-level model object for NoticeContainer.

My concern is that the fields aren't the same for each Notice. Could we do this at the Notice abstract class level?

Reply:
I think you'd have something like:

```java
public class NoticeView {
  private final Notice notice;
  private final JsonObject json; 
  public NoticeView(Notice notice) {
    this.notice = notice;
    this.json = notice.getContext().getAsJsonObject();
  }
  String getNameAsMarkdownId() { return notice.getClass().getSimpleName().toLowerCase(); }
  List<String> getFields() { ... }
  Object getValueForField(String field) { ... }  
}
```

Because Notice is an abstract base class that defines the getContext() method, then you can pass in any time of concrete Notice sub-class and it will still work the same way, regardless of the fields present.  Does that make sense?

Reply:
Oh! Yes, it does. Thank you for the clarification.

Reply:
@bdferris-v2 I added NoticeView.java and NoticeViewTest.java to fix this. Please let me know if it makes more sense.